### ISSUE TITLE
multiple kinect v2 on one machine?

### ISSUE THREAD

**Issue by joshff9** [OpenKinect/libfreenect2#97]
Does anyone know if libfreenect2 supports multiple kinect v2 on one machine? Official SDK does not support this feature, it would be very nice if libfreenect2 support this.

**Comment by avidbot1** [comment]
I've tried multiple devices on 2 computers: a dell xps (intel core i7, nvidia geforce gtx 745, 12Gb RAM), and an intel NUC (intel core i5, intel HD 5000 graphics, 4Gb RAM). Both devices are running Ubuntu 12.04 with linux kernel version 3.16, use USB 3.0 ports for the sensors, and I've changed the `/etc/default/grub` file in both from:

`GRUB_CMDLINE_LINUX_DEFAULT="quiet splash"`
to
`GRUB_CMDLINE_LINUX_DEFAULT="quiet splash usbcore.usbfs_memory_mb=4000"`

I am able to run protonect with multiple devices successfully on the dell xps. On the NUC I can select which sensor to run, and run them individually, but when I try to run a second device when one is already running it crashes with this output:

```
[OpenCLDepthPacketProcessor::listDevice]  devices:
  0: Intel(R) Core(TM) i5-4250U CPU @ 1.30GHz (CPU)[Intel(R) Corporation]
[OpenCLDepthPacketProcessor::init]  selected device: Intel(R) Core(TM) i5-4250U CPU @ 1.30GHz (CPU)[Intel(R) Corporation]
[Freenect2Impl] enumerating devices...
[Freenect2Impl] 15 usb devices connected
[Freenect2Impl] found valid Kinect v2 @3:5 with serial 503614242542
[Freenect2Impl] found valid Kinect v2 @3:4 with serial 500902543042
[Freenect2Impl] found 2 devices
[Freenect2DeviceImpl] opening...
[Freenect2DeviceImpl] opened
[Freenect2DeviceImpl] starting...
[Freenect2DeviceImpl] ReadData0x14 response
92 bytes of raw data
0x0000:  00 00 12 00 00 00 00 00 01 00 00 00 43 c1 1f 41    ............C..A
0x0010:  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00    ................
0x0020:  0a 21 33 55 20 00 17 ba 00 08 00 00 10 00 00 00    .!3U ...........
0x0030:  00 01 00 00 00 10 00 00 00 00 80 00 01 00 00 00    ................
0x0040:  31 33 00 00 00 06 10 08 47 4d 42 30 31 34 2e 31    13......GMB014.1
0x0050:  58 00 00 00 00 00 00 00 00 00 00 00                X...........

[CommandTransaction::receive] bulk transfer failed! libusb error -1: LIBUSB_ERROR_IO
*** glibc detected *** ./Protonect: double free or corruption (top): 0x0000000003efd790 ***
======= Backtrace: =========
/lib/x86_64-linux-gnu/libc.so.6(+0x7e846)[0x7fcc63fd7846]
/home/avidbot/libfreenect2/examples/protonect/lib/libfreenect2.so(_ZN12libfreenect28protocol18CommandTransaction6Result10deallocateEv+0x12)[0x7fcc651de352]
/home/avidbot/libfreenect2/examples/protonect/lib/libfreenect2.so(_ZN12libfreenect28protocol18CommandTransaction6Result8allocateEm+0x22)[0x7fcc651de392]
/home/avidbot/libfreenect2/examples/protonect/lib/libfreenect2.so(_ZN12libfreenect28protocol18CommandTransaction7executeERKNS0_11CommandBaseERNS1_6ResultE+0x39)[0x7fcc651de919]
/home/avidbot/libfreenect2/examples/protonect/lib/libfreenect2.so(_ZN12libfreenect219Freenect2DeviceImpl5startEv+0xb52)[0x7fcc651e1322]
./Protonect(main+0x1bd)[0x401bbd]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xed)[0x7fcc63f7a76d]
./Protonect[0x4022f1]
```

 As far as I can tell I've done everything the same on the dell xps and the NUC so I'm wondering if it's an issue with the hardware. 

Any ideas on how to get this working on the NUC?

**Comment by floe** [comment]
Purely a guess, but since you've set the USB memory pool to 4GB and your NUC also has "only" 4GB of RAM, could simply adding more memory make a difference?

**Comment by avidbot1** [comment]
When running 1 instance of protonect, 'top' lists Protonect memory at 5.1%. 

Here's the output of free -m with 1 instance of Protonect running:
```
             total       used       free     shared    buffers     cached
Mem:          3881       3229        651          0        214       2021
-/+ buffers/cache:        992       2888
Swap:         4022          0       4022
```
And here's the output with nothing running:
```
             total       used       free     shared    buffers     cached
Mem:          3881       3068        812          0        214       2021
-/+ buffers/cache:        833       3048
Swap:         4022          0       4022
```
From this it looks like Protonect isn't actually using that much memory, which makes it seem like adding RAM wouldn't make any difference. Is this an accurate representation of my memory usage?

**Comment by avidbot1** [comment]
Not sure how useful this is, but I traced the LIBUSB_ERROR_IO back to submit_bulk_transfer in linux_usbfs.c.

    r = ioctl(dpriv->fd, IOCTL_USBFS_SUBMITURB, urb);

returns -1.

**Comment by avidbot1** [comment]
I've upgraded to 10Gb of RAM and I'm still getting the same error, so it seems to be more than just a RAM issue.

**Comment by mu-777** [comment]
Hi, I have the same error when using 2 kinects at the same time on one computer.
Does noone fix it still?

My computer has intel i7-5820K, GeForce GTX 980 and 16GB RAM under ubuntu 15.04 64bit.
I think this is quite enough to work 2 kinects but I have the the message
```
[CommandTransaction::receive] bulk transfer failed! libusb error -1: LIBUSB_ERROR_IO
```
and it doesn't work.

Someone helps me, thanks.

**Comment by josephduchesne** [comment]
I had this issue and resolved it on a Lenovo Y-50 by changing:
GRUB_CMDLINE_LINUX_DEFAULT="quiet splash"
to
GRUB_CMDLINE_LINUX_DEFAULT="quiet splash usbcore.usbfs_memory_mb=64"
in /etc/default/grub
then running `sudo update-grub`

**Comment by stfuchs** [comment]
Thanks, I had the same issue and your solution worked fine for me.

**Comment by floe** [comment]
Added to the README.

**Comment by xlz** [comment]
Does `echo 64 > /sys/module/usbcore/parameters/usbfs_memory_mb` (root) work? I think this runtime method would be friendlier than editing kernel parameters.

**Comment by siddharthachandra** [comment]
Hi, could anyone specify the hardware setup that was used to exploit the 5 kinect sensors simultaneously? I'd especially like to know which motherboard was used. Thanks in advance.

**Comment by xlz** [comment]
@siddharthachandra On the motherboard each Kinect needs 3Gbps PCI-e bandwidth. Double that for safety margin, that's 30Gbps. There are probably some more bandwidth considerations for USB 3.0 extension cards depending on the link layout.

**Comment by SteveSmithStyku** [comment]
I am attempting to use two kinect v2 sensors on a windows 8.1 machine using libfreenect. Each sensor is connected to its own USB 3.0 pcie card, and each card connected to an x8 pcie slot. I have confirmed that each sensor individually runs perfectly with Microsoft Kinect SDK at 30 fps on each card.

I have modified the Protonect example to only show depth and color. when I do this, I seem to get an acceptable framerate for a single sensor. However, when I try to use a second sensor at the same time, I seem to get a similar error as avidbot1, mu-777, josephduchesne, and stfuchs:

[Error] [protocol::CommandTransaction] bulk transfer failed: LIBUSB_ERROR_IO Input/Output Error

And I am perpetually stuck on SyncMultiFrameListener::waitForNewFrame.

Has anyone managed to get two kinect v2s working on windows? Does anyone know what the equivalent fix on windows to the above linux fix of setting usbfs memory to 64 mb is?

**Comment by xlz** [comment]
> I have modified the Protonect example

Well you don't have to do this? There are `-norgb | -nodepth` options.

LIBUSB_ERROR_IO is very general. The low-level error can only be revealed by debugging with LIBUSB_DEBUG=4. That is required to determine what kind of low-level error that was.

**Comment by SteveSmithStyku** [comment]
@xlz thank you for the quick reply!

I needed to modify the protonect example because I need to use 2 devices, not 1. Sorry for the confusion.

The debugging output is very VERY long, so I'll truncate it at the point where I get "[protocol::CommandTransaction] bulk transfer failed: LIBUSB_ERROR_IO Input/Output Error"

Please see the attached file
[libusb_debug.txt](https://github.com/OpenKinect/libfreenect2/files/147145/libusb_debug.txt)
(backup: http://pastebin.com/raw/QGCMxbRD)

And the output without LIBUSB_DEBUG defined:

[Info] [Freenect2Impl] enumerating devices...
[Info] [Freenect2Impl] 10 usb devices connected
[Info] [Freenect2Impl] found valid Kinect v2 @24:2 with serial 030260350347
[Info] [Freenect2Impl] found valid Kinect v2 @32:2 with serial 181369735247
[Info] [Freenect2Impl] found 2 devices
[Info] [Freenect2DeviceImpl] opening...
[Info] [Freenect2DeviceImpl] opened
[Info] [Freenect2DeviceImpl] opening...
[Info] [Freenect2DeviceImpl] opened
[Info] [Freenect2DeviceImpl] starting...
[Info] [Freenect2DeviceImpl] submitting rgb transfers...
[Info] [Freenect2DeviceImpl] submitting depth transfers...
[Info] [Freenect2DeviceImpl] started
[Info] [Freenect2DeviceImpl] starting...
[Error] [protocol::CommandTransaction] bulk transfer failed: LIBUSB_ERROR_IO Input/Output Error
[Info] [DepthPacketStreamParser] 46 packets were lost
[Info] [DepthPacketStreamParser] 1 packets were lost
[Info] [DepthPacketStreamParser] 1 packets were lost
[Info] [OpenGLDepthPacketProcessor] avg. time: 24.7773ms -> ~40.3595Hz
[Info] [OpenGLDepthPacketProcessor] avg. time: 13.2577ms -> ~75.4279Hz
[Info] [TurboJpegRgbPacketProcessor] avg. time: 24.9527ms -> ~40.0758Hz
(this continues in this fashion after the error)

Please let me know if I need to create a separate issue, thanks.

**Comment by xlz** [comment]
Are you using two Kinects within one program? Can you try to run two instances of `Protonect.exe 030260350347` and `Protonect.exe 181369735247`?

The log shows
```
[ 2.017330] [00000f18] libusb: debug [handle_events] poll fds modified, reallocating
[ 2.017330] [00000f18] libusb: debug [handle_events] poll() 82 fds with timeout in 983ms
[ 2.017330] [00000f18] libusb: debug [handle_events] poll() returned -1
[ 2.017330] [00000f18] libusb: error [handle_events] poll failed -1 err=5
[ 2.017330] [00000f18] libusb: error [sync_transfer_wait_for_completion] libusb_handle_events failed: LIBUSB_ERROR_IO, cancelling transfer and retrying
```
It's probably because it was polling too many fds (https://marc.info/?l=libusb-devel&m=144933065117809&w=2).

**Comment by SteveSmithStyku** [comment]
@xlz By passing in the serial number, yes, I am able to run each device individually.

Regarding "polling too many fds," is there a way to control this?

It may be helpful for you to see the exact test code I am running (which is very simple, just to make sure things work), maybe I am doing something wrong? I used https://openkinect.github.io/libfreenect2/ as a guide and reused the Viewer class in the Protonect example.

[Protonect_modified.cpp.txt](https://github.com/OpenKinect/libfreenect2/files/147181/Protonect_modified.cpp.txt)

Thanks again for your continued help.

**Comment by xlz** [comment]
@SteveSmithStyku your bug to be followed and fixed in #593.
